zcash_client_sqlite/src/lib.rs

```diff
@@ -24,24 +24,545 @@
+        ])?;
+
+        Ok(())
+    }
+
+    fn rewind_to_height<P: consensus::Parameters>(
+        &mut self,
+        parameters: &P,
+        block_height: BlockHeight,
+    ) -> Result<(), Self::Error> {
+        wallet::rewind_to_height(self.conn, parameters, block_height)
+    }
+
+    fn put_tx_meta(
+        &mut self,
+        tx: &WalletTx,
+        height: BlockHeight,
+    ) -> Result<Self::TxRef, Self::Error> {
+        let txid = tx.txid.0.to_vec();
+        if self.stmt_update_tx_meta.execute(&[
+            u32::from(height).to_sql()?,
+            (tx.index as i64).to_sql()?,
+            txid.to_sql()?,
+        ])? == 0
+        {
+            // It isn't there, so insert our transaction into the database.
+            self.stmt_insert_tx_meta.execute(&[
+                txid.to_sql()?,
+                u32::from(height).to_sql()?,
+                (tx.index as i64).to_sql()?,
+            ])?;
+
+            Ok(self.conn.0.last_insert_rowid())
+        } else {
+            // It was there, so grab its row number.
+            self.stmt_select_tx_ref
+                .query_row(&[txid], |row| row.get(0))
+                .map_err(SqliteClientError::from)
+        }
+    }
+
+    fn put_tx_data(
+        &mut self,
+        tx: &Transaction,
+        created_at: Option<time::OffsetDateTime>,
+    ) -> Result<Self::TxRef, Self::Error> {
+        let txid = tx.txid().0.to_vec();
+
+        let mut raw_tx = vec![];
+        tx.write(&mut raw_tx)?;
+
+        if self.stmt_update_tx_data.execute(&[
+            u32::from(tx.expiry_height).to_sql()?,
+            raw_tx.to_sql()?,
+            txid.to_sql()?,
+        ])? == 0
+        {
+            // It isn't there, so insert our transaction into the database.
+            self.stmt_insert_tx_data.execute(&[
+                txid.to_sql()?,
+                created_at.to_sql()?,
+                u32::from(tx.expiry_height).to_sql()?,
+                raw_tx.to_sql()?,
+            ])?;
+
+            Ok(self.conn.0.last_insert_rowid())
+        } else {
+            // It was there, so grab its row number.
+            self.stmt_select_tx_ref
+                .query_row(&[txid], |row| row.get(0))
+                .map_err(SqliteClientError::from)
+        }
+    }
+
+    fn mark_spent(&mut self, tx_ref: Self::TxRef, nf: &[u8]) -> Result<(), Self::Error> {
+        self.stmt_mark_recived_note_spent
+            .execute(&[tx_ref.to_sql()?, nf.to_sql()?])?;
+        Ok(())
+    }
+
+    // Assumptions:
+    // - A transaction will not contain more than 2^63 shielded outputs.
+    // - A note value will never exceed 2^63 zatoshis.
+    fn put_received_note<T: ShieldedOutput>(
+        &mut self,
+        output: &T,
+        nf: Option<&[u8]>,
+        tx_ref: Self::TxRef,
+    ) -> Result<Self::NoteRef, Self::Error> {
+        let rcm = output.note().rcm().to_repr();
+        let account = output.account().0 as i64;
+        let diversifier = output.to().diversifier().0.to_vec();
```

Comment:
This allocation shouldn't be necessary? Or is it part of the requirements for `rusqlite` named arguments?

Reply:
Yup, rusqlite needs a vec.